src/Reflection/ReflectionClass.php

```diff
@@ -360,27 +360,35 @@ private function createMethodsFromTrait(ReflectionMethod $method): array
         return $methods;
     }
 
-    /** @return list<ReflectionMethod> */
-    private function getParentMethods(): array
+    /**
+     * @param list<class-string> $classNameStack
+     *
+     * @return list<ReflectionMethod>
+     */
+    private function getParentMethods(array &$classNameStack = []): array
```

Comment:
holy guacamole by-ref stack passing: that is a lot of methods with by-ref added :O

Reply:
yeah :/ I could reduce them a bit at least

Reply:
Oh, sorry, it was more of a comment on the complexity: I think you may actually have broken it, by removing the by-ref in 04cfc28

Reply:
no worries, you're right. interestingly tests are still green (I just added more to be sure) but indeed those should be refs. I kind of removed them too quickly because that fixed psalm errors.. psalm seems to be "loosing" the type when adding e.g. `use (&$classNameStack)` to closures. I'm not used to psalm, can you quickly point me to something that explains this maybe?

Reply:
Ok, so by-ref types need a `@psalm-param-out` for this sort of stuff, AFAIK.

An alternative to this design can be achieved by having a small mutable `@internal` data structure: that removes `&` from all this code :)

Reply:
hmm I need to think a bit more about this. It can get messy quite fast, in theory I'd need to convert multiple arrow to anonymous functions, just to use by-ref variables inside. Also in e.g. https://github.com/Roave/BetterReflection/pull/1241/files#diff-c58f24d134490b09a18aa65366bc9ab2d5bf2dbf493bea9f12b365807a5b5e22R1628 I guess.

Reply:
My 2 cents - I really dislike references for this purpose, a property is much more clear: https://github.com/phpstan/phpstan-src/commit/33771e5d8ee8b6dbae4f42ce3f8b424a81f68dc9

Reply:
the small mutable internal data structure sounds like a good combination of both worlds. but in the simplest form it would just be an array property I guess? :) Or should I create something like a simple immutable ClassNameStack VO maybe?

Reply:
An array property suffices 👍

Reply:
I was not thinking this through enough. The property on `ClassReflection` won't be of much help, since the code is creating new ClassReflection instances when recursing which would reset the property, except I preserve this somehow. The use-case is a bit different to what Ondrej posted I think. 🤔
I pushed a commit that passes around a simple immutable object instead, but it feels just like working-around the by-ref thing, doesn't it? 😅 Sorry that this turned out more complex than initially thought.

Reply:
It is working around the by-ref, but this design is much cleaner, and exactly what was suggested above :+1: